### ISSUE TITLE
Modular Configure

### ISSUE THREAD

**Issue by scolapasta** [IQSS/dataverse#4230]
We need the ability to configure some external tools - in this case Psi for differential privacy. We would like to do this in a modular way so that if some other installation wants to create other tools to be configured, they can.

By "configure" we mean tools that are meant to be run by the data "owner" or other such authorized user. (the other similar concept is "Modular Explore" #3657, which is external tools meant to be run by other users).

**Comment by djbrooke** [comment]
We should consider that tools will have:

- Text to display on the button
- Information in the popup
- Link to the tool
- Things (files, metadata, etc.) that need to be passed to the tool

https://docs.google.com/presentation/d/1rPpaHNW-5QIskYvn_4afzXzU_wcdkZb2SjEDhpUhhLs/edit#slide=id.g26f2e4bd68_0_36

**Comment by pdurbin** [comment]
@scolapasta please take a look at 6fc769c and let me know if you think this is a good reflection of what we talked about this afternoon. I'm happy to make changes. Thanks!

**Comment by pdurbin** [comment]
The definition of done of this issue is not clear. I went ahead and made pull request #4243 and I'm moving it to Code Review.

In the GUI, you can now see dynamic tools like this after doing the setup described in the Installation Guide:

![screen shot 2017-10-27 at 11 14 49 am](https://user-images.githubusercontent.com/21006/32111575-daa854e4-bb08-11e7-8559-7e71b78188cf.png)

There's some chatter in the Google Presentation linked above about the button:

![screen shot 2017-10-27 at 11 21 00 am](https://user-images.githubusercontent.com/21006/32111646-19a70730-bb09-11e7-9f1b-e8bf562eeb41.png)

**Comment by pdurbin** [comment]
> In the GUI, you can now see dynamic tools like this after doing the setup described in the Installation Guide

I just talked through this @mheppler and he gets that for now I just wanted to put something in the UI but that we plan to iterate on the design. I mentioned that @kevinworthington and @amberleahey might be interested in this new `externaltool` table or whatever we end up calling it. Tools like Data Explorer (#2216 but a new issue would probably be better) may appear in a table like this some day. What needs to be worked out is what parameters each tool needs and how to send them (as query parameters? as headers? a mix of both?). For now in the database I'm just storing a JSON object for extreme flexibility of what the tool needs.

**Comment by kevinworthington** [comment]
It would definitely help to have access to properties relating to privacy/security within the Data Explorer. Right now the system operates as a standalone application which makes calls to three Dataverse APIs to run (DDI, prep file and tab file endpoints).  Having properties passed through header variables would work, though to maintain modularity, there could be another call from the  Data Explorer to retrieve these privacy/security properties as a json file.

**Comment by pdurbin** [comment]
@scolapasta and I talked about this today:

![img_20171102_142927](https://user-images.githubusercontent.com/21006/32347123-0d007872-bfe6-11e7-9700-4f1ca8f56ba2.jpg)

**Comment by scolapasta** [comment]
From the image:

Besides the more obvious dynamic text and links we need for external tools, a couple of aspects that we need to handle in a modular way are:
- When to show the links to the external tool
- How to handle the link to the external tool (i.e. wha dynamic info to send)

In both these cases, we can think of building these in incremental ways, with each step building on the previous step, building from simplest, but supporting fewer tools, to most complex, but also most flexible.

For when to show the link, we have 3 steps:
A. simple list of tools, will only support one operation (i.e. Configure)
B. Add "Operation" to list, in order to support multiple operations (i.e. Configure vs. Explore)
C. Add Subject of Operation" in order to support different operation on different kind of operations (i.e. tools that work on tabular files, like DataExplorer, vs tools that work on geospatial, like World Map.

For how to handle link we similarly have 3 steps:
1. Dictate the exact parameter names and values that an external tool will need.
2. Define parameters dynamically from a list of "reserved" terms that we define.
3. Define individual "Handler" Java classes and SPI interface, in order to handle any tool.

In order to support Modular Configure for PSI (which is the goal of this issue, we need to support the most simple, i.e. A1). This would allow an admin to configure PSI or other configure tool that uses the same parameters (name and value) dynamically.

Therefore, the definition of done for this issue is A1, though we may already start supporting a more complex level, on our path to support #3657 (i.e. support A2).

**Comment by pdurbin** [comment]
@scolapasta thanks, nice summary.

Note to self that @mheppler mentioned that #4233 is dependent on this issue: https://github.com/IQSS/dataverse/issues/4233#issuecomment-340870897

@kevinworthington either here or at #3657 I'm probably to throw some JSON at you and ask if you think it describes well enough the parameters that Data Explorer (#4249) needs. See `psi.json` at 6fc769c for the direction I'm going.

**Comment by kevinworthington** [comment]
Hi Phil,
Definately looking forward to seeing this piece of json you mentioned. I'll
be in the office tomorrow and plan on attending the call if you'd like to
discuss things further.

Kevin

On Nov 2, 2017 8:03 PM, "Philip Durbin" <notifications@github.com> wrote:

@scolapasta <https://github.com/scolapasta> thanks, nice summary.

Note to self that @mheppler <https://github.com/mheppler> mentioned that
#4233 <https://github.com/IQSS/dataverse/issues/4233> is dependent on this
issue: #4233 (comment)
<https://github.com/IQSS/dataverse/issues/4233#issuecomment-340870897>

@kevinworthington <https://github.com/kevinworthington> either here or at
#3657 <https://github.com/IQSS/dataverse/issues/3657> I'm probably to throw
some JSON at you and ask if you think it describes well enough the
parameters that Data Explorer (#4249
<https://github.com/IQSS/dataverse/issues/4249>) needs. See psi.json at
6fc769c
<https://github.com/IQSS/dataverse/commit/6fc769ccb4c2aca4b5e919ac99353977fb04e890>
for the direction I'm going.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/IQSS/dataverse/issues/4230#issuecomment-341593380>, or mute
the thread

**Comment by pdurbin** [comment]
> I'll be in the office tomorrow and plan on attending the call if you'd like to discuss things further.

@kevinworthington great! Officially tomorrow's community call is about internationalization ( https://groups.google.com/d/msg/dataverse-community/d7BnFiRA3_I/ggFynEZvBwAJ ) but if you call it and bring up modularity I'm happy to talk about what I'm up to in the `4230-modular-configure` branch.

**Comment by pdurbin** [comment]
@kevinworthington we only touched on external tools in the community call today but I wanted to touch base with your about what I've done so far. As of ff74386 here's some JSON I'm suggesting to describe an external tool:

```
{
  "displayName": "Privacy-Preserving Data Preview",
  "description": "The link below goes to a tool that helps data depositors release privacy preserving statistics describing a dataset.\n\nThis is done by distributing a privacy budget across different possible statistical calculations. This tool is called the budgeter. Users select which statistics they would like to calculate and are given estimates of how accurately each statistic can be computed. They can also redistribute their privacy budget according to which statistics they think are most valuable in their dataset.",
  "toolUrl": "https://beta.dataverse.org/custom/DifferentialPrivacyPrototype/UI/code/interface.html",
  "toolParameters": {
    "queryParameters": [
      {
        "fileid": "{fileId}"
      },
      {
        "key": "{apiToken}"
      }
    ]
  }
}
```

This is still very much in flux but the idea is that the `displayName` appears on the button and the `description` appears somewhere in the UI to explain what the tool is used for. For now `toolParameters` only supports `queryParameters`, which I **think** is all that Data Explorer (#4249) needs. The funny `{fileId}` and `{apiToken}` fields are reserved words that dynamically insert a file id or a user's API token. It looks like Data Explorer will want a `uri` query parameter that also includes the "site URL" for a Dataverse installation plus a file id ( https://scholarsportal.github.io/Dataverse-Data-Explorer/?uri=https://dataverse.scholarsportal.info/api/access/datafile/8988/metadata/ddi ) so I guess I'll need to add some more logic to the parser to handle more complicated query parameters, maybe like this:

    "queryParameters": [
      {
        "uri": "{siteUrl}/api/access/datafile/{fileId}/metadata/ddi"
      },

I hope this is making sense! Also, I'd love to hear more of your thoughts on how the JSON file above would look in both English and French (extensible to other languages) Over in the notes at https://groups.google.com/d/msg/dataverse-community/d7BnFiRA3_I/O6cYqSBfCAAJ I attempted to summarize what you were saying at "different languages as keys in JSON files" but if you could give me a more concrete example using Data Explorer, I would really appreciate it!

**Comment by pdurbin** [comment]
@scolapasta I just moved this into Code Review at https://waffle.io/IQSS/dataverse for you to take a look.

**Comment by kevinworthington** [comment]
Hi @pdurbin,
I can totally work with the json file you have described once the siteUrl has been added. Having support for a localization variable would be fantastic, though I could also use a cookie. 

The Data Explorer supports localization by having all messaging in an external json file loaded at run-time. The variable references for these messages are embedded in the UI and the Angular framework used handles the swapping these references with their value. Depending on the set language a different external json file is loaded translating all the messaging.

**Comment by pdurbin** [comment]
> I can totally work with the json file you have described once the siteUrl has been added.

@kevinworthington great! Please be advised that adding that "site URL" reserved word will probably be deferred until Modular Explore (#3657). The names of these issues are someone confused (as discussed at standup this morning) but your Data Explorer tool is an "explore" tool and [PSI](https://privacytools.seas.harvard.edu/psi) is a "configure" tool in the sense that it's not read only.

> Having support for a localization variable would be fantastic, though I could also use a cookie.

Does this mean that I we should not worry about internationalization for these external tools for now? You have a workaround? You'll use a cookie?

**Comment by scolapasta** [comment]
Kevin ( @kevinworthington correct me if I'm wrong) is only referring to the internationalization on the  external tool side (in this case Data Explorer) which we shouldn't have to worry about, unless it's something we need to pass).

We'll always have to handle the internationalization for any text on the dataverse side (e.g. name of explore tool).

**Comment by pdurbin** [comment]
Right, to clarify, my concern is showing "Data Explore" in French on the button on the Dataverse side, @kevinworthington . I'm thinking the JSON example above should have a key for `fr` for `displayName` and `description`, for example. The button on the Dataverse side uses `displayName` and we plan to show the description of external tools within the Dataverse GUI as well. Here's a screenshot of the external tool "PSI" showing a description from #4233

![32343315-75db5d6a-bfd9-11e7-8d83-1ca7ef1a47d5](https://user-images.githubusercontent.com/21006/32524974-54c0ffc0-c3f0-11e7-80d6-4604149534dc.png)

**Comment by kevinworthington** [comment]
Hi @pdurbin and @scolapasta,
There's no need to pass any language information to the Data Explorer at present. 
And it might be best to wait on adding translated text options until there's a plan for multilingual support.

**Comment by pdurbin** [comment]
> it might be best to wait on adding translated text options until there's a plan for multilingual support

Agreed. For now I'm storing `displayName` and `description` in the new `externaltool` table in English. I left a TODO on the entity class indicating that some day we should think about internationalization. Whoops, I spelled it wrong: https://github.com/IQSS/dataverse/blob/8c697330f9874f942dbe70e91e27304510bfefea/src/main/java/edu/harvard/iq/dataverse/externaltools/ExternalTool.java#L22

**Comment by dlmurphy** [comment]
Just talked with @scolapasta about documentation for this issue. We're interested in pursuing a new organizational strategy for the Installation Guide where external tools that can be integrated into Dataverse (e.g. PSI) are grouped together on one page. Each tool will have its own section of that page that will introduce it conceptually and link out to whatever documentation the creators of the tool have put together.

**Comment by pdurbin** [comment]
@dlmurphy cool. I'm done with the `4230-modular-configure` branch for the night. You can push your changes there. This is the branch we intend to put through QA.

**Comment by pdurbin** [comment]
@scolapasta I think I addresses all the changes you wanted so I'm moving this back to code review for you to take a look.

**Comment by pdurbin** [comment]
@dlmurphy I took a look at 0e41588 and I'm wondering if you and @scolapasta talked about if we want to name PSI specifically as an external tool in the Installation Guide. I was considering re-writing `psi.json` to be `awesomeTool.json` or whatever and just generically describe how to set up any external tool. The background on this is that we want to encourage external tools to write their own documentation. That's what Gustavo and talked about, anyway, and I think it's a good idea.

**Comment by dlmurphy** [comment]
@pdurbin, when I spoke with @scolapasta yesterday we figured the External Tools page could be sort of a roster of different external tools, with brief introductions explaining them on a conceptual level, any very basic info needed to install the tool in Dataverse, and a link to the tool's website/contact info. My goal here is not to write an exhaustive guide to PSI, but rather to explain only what is necessary for Dataverse admins to decide whether they want to enable the tool or not, and give them a link to where to go for more info than that.

**Comment by pdurbin** [comment]
@dlmurphy ok, in that case I think we should think about promoting other tools that can be integrated with Dataverse. If you look at https://dataverse.org/integrations you can see that the following tools integrate with Dataverse:

- Two Ravens
- World Map
- Open Science Framework
- Open Journal Systems
- RSpace
- SHARE

The use story is, "As someone who just installed Dataverse (or asked someone to install it for me), I'd like to know what integrations are available to me now that I have a new shiny Dataverse toy to play with."

There's also a list at http://guides.dataverse.org/en/4.8.2/api/apps.html but it makes more sense to me to put the list in the Installation Guide.

There might be more integrations I'm forgetting. The point is that people may not even know they are possible.

**Comment by dlmurphy** [comment]
That's a really helpful list! I do think that this page in the docs should absolutely include all of those tools, but I think trying to include them all right now might count as scope creep for this issue. What do you think @scolapasta? We could always create a new issue for putting all of those tools into this doc page.

**Comment by pdurbin** [comment]
I forgot to mention that I floated this idea past @mercecrosas months ago and she liked it. I just haven't gotten around to creating an issue. This issue makes it suddenly feel on-topic.

**Comment by pdurbin** [comment]
@scolapasta @dlmurphy and I discussed https://github.com/IQSS/dataverse/blob/0e41588e8dcda2498bfab810195e55502d93da35/doc/sphinx-guides/source/installation/external-tools.rst after standup and this is my current understanding of the docs we've written so far:

- We don't want to lump all the integrations I mentioned above into the new "External Tools" page. I will make an issue about a new page in the Installation Guide for integrations after I give the community some time to reply to the post I just made about integrations: https://groups.google.com/d/msg/dataverse-community/tDb8WKWFnqw/ySOeBNeIAQAJ
- We are on the fence about where we should continue to mention PSI specifically on the "External Tools" page. I offered to rename `psi.json` to `awesomeTool.json` and generalize all the text to make this page about supporting external tools in general and not PSI specifically.

**Comment by pdurbin** [comment]
> I will make an issue about a new page in the Installation Guide for integrations

Done: #4283

**Comment by pdurbin** [comment]
@scolapasta and I met and he requested changes, which I implemented in 2de0187.

**Comment by pdurbin** [comment]
@scolapasta and I considering making a change like this:

```
-    private String getQueryParam(String key, String value) {
+    private String getQueryParam(String key, String value) throws Exception {
         DataFile df = getDataFile();
         if (df == null) {
             logger.fine("DataFile was null!");
@@ -104,11 +103,11 @@ public class ExternalToolHandler {
             case "{apiToken}":
                 return key + "=" + apiTokenString;
             default:
-                return key + "=" + value;
+                throw new Exception();
         }
     }
```

Because this is a checked exception, this changes a fair amount of code. Also, we might make our own exception (`UnsupportedReservedWordException` perhaps) or try to use https://docs.oracle.com/javase/8/docs/api/java/text/ParseException.html

Basically, the idea is to start out more strict and only allow reserved words as parameters.

**Comment by pdurbin** [comment]
@scolapasta I'm ready for more code review as of 6e3098e . The I ended up using a RunTimeException to make the change smaller.

**Comment by pdurbin** [comment]
New todo list based on code review as of 6e3098e 

- [ ] Let's make this error message more generic, Unknown reserved word...
- [ ] I'd rather it be a checked exception. ParseException didn't seem to work because of the offset, but then how about IllegalArgument. That is, if we don't want to make our own. But I'd be fine with IllegalArguement
- [ ] Unless I missed it, I don't see this being checked on the add external tool, which means an invalid tool definition can get into db.

**Comment by pdurbin** [comment]
@scolapasta I moved this to code review. As we discussed with @matthew-a-dunlap this morning, I made the fixes in the `4230-4233-external-tool-psi` branch. There's no pull request but you can do the code review with the "compare" feature of GitHub:

https://github.com/IQSS/dataverse/compare/4230-4233-external-tool-psi

**Comment by pdurbin** [comment]
More todo after code review as of f4bf37e

- [ ] make enum values just `fileId` and `apiToken` and make `{` and`}` constants

Other stuff, to be discussed.

**Comment by pdurbin** [comment]
@scolapasta I addressed what I see as the two most pressing items in your latest round of code review. See the latest todo list above. I hope we're close. Please advise.

**Comment by matthew-a-dunlap** [comment]
Here is the list of code review feedback on the UI side of modular configure:

- [ ] The logic for what tools to show for a file should be moved out of the UI and instead be at creation of the list of handlers for the files. 
    * Eventually we will likely add the fileto the configurations themselves. At the moment we will only have infrastructure to write back in specific cases (e.g. psi fuzzy summary statistics) so file typing is less important.
- [ ] Even if the handler list is well populated, passing the full list up front does not work the UI from a lazy loading perspective
- [ ] Make sure popup matches map logic on when to show buttons when dataverse is a widget (e.g. shown in an iframe)

**Comment by pdurbin** [comment]
New todo list based on code review as of ca7ab99

- move `toJson` from ExternalToolHandler.java to ExternalTool.java
- remove "test" from `"/api/admin/externalTools/test/file/" + fileId` API endpoint

**Comment by pdurbin** [comment]
@scolapasta ok, I did both items. Moved to code review.

**Comment by scolapasta** [comment]
Looking good on changes except:
[ ]  the name of the method still references externalToolHandler
[ ] some no longer needed code about APIs (lines 65-67)

**Comment by landreev** [comment]
Synced up the branch with develop. (the merge conflict on the DatasetPage was trivial)
This is in preparation for creating a mega-super-branch spanning #4230, #4233, #4231, #4232 and #4307. 
The goal is to install this combined branch on beta today, for the PSI demo testing work. 
Please note that for the purposes of QA, the 2 branches (4230-4233-external-tool-psi and 4231-psi-metadata-infrastracture) will stay separate.

**Comment by pdurbin** [comment]
@kcondon the main entry point for what this issue about is the new "External Tools" page in the Admin Guide: https://github.com/IQSS/dataverse/blob/8b9e125b2a4247162f0d760b535c5c5a08420894/doc/sphinx-guides/source/installation/external-tools.rst
